packages/flutter/test/widgets/nested_scroll_view_test.dart

```diff
@@ -676,8 +676,309 @@ void main() {
+      final GlobalKey<NestedScrollViewState> globalKey = GlobalKey();
+
+      await tester.pumpWidget(NestedScrollViewWithGlobalKey(globalKey: globalKey));
+
+      expect(globalKey.currentState.outerController.position.pixels, 0);
+      expect(globalKey.currentState.innerController.position.pixels, 0);
+
+      const double innerPosition = 0e0, outerPosition = 0e0;
+
+      testModifyingInnerController(globalKey, innerPosition, outerPosition);
+      testModifyingOuterController(globalKey, innerPosition, outerPosition);
+    });
+
+    testWidgets('inner body scrolled, header not scrolled', (WidgetTester tester) async {
+      final GlobalKey<NestedScrollViewState> globalKey = GlobalKey();
+
+      await tester.pumpWidget(NestedScrollViewWithGlobalKey(globalKey: globalKey));
+
+      expect(globalKey.currentState.outerController.position.pixels, 0);
+      expect(globalKey.currentState.innerController.position.pixels, 0);
+
+      const double innerPosition = 1e3, outerPosition = 0e0;
+
+      testModifyingInnerController(globalKey, innerPosition, outerPosition);
+      testModifyingOuterController(globalKey, innerPosition, outerPosition);
+    });
+
+    testWidgets('inner body not scrolled, header partially scrolled', (WidgetTester tester) async {
+      final GlobalKey<NestedScrollViewState> globalKey = GlobalKey();
+
+      await tester.pumpWidget(NestedScrollViewWithGlobalKey(globalKey: globalKey));
+
+      expect(globalKey.currentState.outerController.position.pixels, 0);
+      expect(globalKey.currentState.innerController.position.pixels, 0);
+
+      const double innerPosition = 0e3, outerPosition = kToolbarHeight / 3;
+
+      testModifyingInnerController(globalKey, innerPosition, outerPosition);
+      testModifyingOuterController(globalKey, innerPosition, outerPosition);
+    });
+
+    testWidgets('inner body scrolled, header partially scrolled', (WidgetTester tester) async {
+      final GlobalKey<NestedScrollViewState> globalKey = GlobalKey();
+
+      await tester.pumpWidget(NestedScrollViewWithGlobalKey(globalKey: globalKey));
+
+      expect(globalKey.currentState.outerController.position.pixels, 0);
+      expect(globalKey.currentState.innerController.position.pixels, 0);
+
+      const double innerPosition = 1e3, outerPosition = kToolbarHeight / 3;
+
+      testModifyingInnerController(globalKey, innerPosition, outerPosition);
+      testModifyingOuterController(globalKey, innerPosition, outerPosition);
+    });
+  });
 }
 
+/// Modifying the inner controller should always push the outer controller
+/// to maxScrollExtent unless setPixels is used.
+void testModifyingInnerController(
+    GlobalKey<NestedScrollViewState> globalKey,
+    double innerPosition,
+    double outerPosition,
+    ) {
+  globalKey.currentState.innerController.position.setPixels(innerPosition);
+  globalKey.currentState.outerController.position.setPixels(outerPosition);
+
+  globalKey.currentState.innerController.jumpTo(5e2);
```

Comment:
Looks like you're missing a test where you change the innerController by less then the kToolbarHeight?